# 03 Trend Across Country

## import modules and dataset

In [23]:
import pandas as pd
import altair as alt
# alt.renderers.enable('jupyterlab')

In [24]:
trials_clean = pd.read_csv('./export/trials_clean.csv')

## Q2: Which country has the most witch trial cases in total?

### Witch Trials By Country - bar chart
Next, we can look at which country has the largest number of witch trials

In [25]:
# Group by country 
trials_by_country = trials_clean.groupby([pd.Grouper(key='gadm.adm0')]).count()[['tried']].reset_index()

# rename the columns
trials_by_country.columns = ['country', 'trials_count']
trials_by_country

country  trials_count
0          Austria            16
1          Belgium           671
2   Czech Republic             1
3          Denmark            90
4          Estonia            17
5          Finland            17
6           France           807
7          Germany          3417
8          Hungary            26
9          Ireland             4
10           Italy           107
11      Luxembourg            20
12     Netherlands           314
13          Norway            20
14          Poland             9
15           Spain            29
16          Sweden           353
17     Switzerland          1272
18  United Kingdom          3750

In [26]:
# chart by altair
alt.Chart(trials_by_country).mark_bar().encode(
    x=alt.X('country'),
    y=alt.Y('trials_count'),
).properties(
    title={
        'text': 'Witch Trials by country',
    }
)

/Users/direwolfsummer/.pyenv/versions/3.9.4/envs/j233-files-3.9.4/lib/python3.9/site-packages/altair/utils/core.py:317: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for col_name, dtype in df.dtypes.iteritems():


<VegaLite 4 object>

If you see this message, it means the renderer has not been properly enabled
for the frontend that you are using. For more information, see
https://altair-viz.github.io/user_guide/troubleshooting.html


![witch trials by country](./img/witch_trials_by_country.png)

United Kingdom and Germany have the most witch trials by far. Together they took up 43.88% of the trials in this dataset. Let's take a closer look.

### Witch Trials in Germany

In [28]:
trials_germany = trials_clean.loc[trials_clean['gadm.adm0']=='Germany']
trials_germany

year  decade  century  tried  deaths            city gadm.adm2  \
539    1500.0    1500     1500      2     NaN  Isny im Allgau  Freiburg   
540    1578.0    1570     1500      1     NaN  Isny im Allgau  Freiburg   
541    1590.0    1590     1500      6     6.0       Erolzheim  Freiburg   
542    1630.0    1630     1600      1     NaN      Heidelberg  Freiburg   
543    1640.0    1640     1600      1     NaN      Heidelberg  Freiburg   
...       ...     ...      ...    ...     ...             ...       ...   
10935  1740.0    1740     1700      2     0.0             NaN       NaN   
10936  1750.0    1750     1700      1     0.0             NaN       NaN   
10937  1760.0    1760     1700      4     0.0             NaN       NaN   
10938  1770.0    1770     1700      6     0.0             NaN       NaN   
10939  1790.0    1790     1700      1     0.0             NaN       NaN   

               gadm.adm1 gadm.adm0       lon      lat     record.source  
539    Baden-Wurttemberg   Germany  10.04200  47.6950  Behringer (1987)  
540    Baden-Wurttemberg   Germany  10.04200  47.6950  Behringer (1987)  
541    Baden-Wurttemberg   Germany  10.07080  48.0917  Behringer (1987)  
542    Baden-Wurttemberg   Germany   8.67243  49.3988  Behringer (1987)  
543    Baden-Wurttemberg   Germany   8.67243  49.3988  Behringer (1987)  
...                  ...       ...       ...      ...               ...  
10935            Sachsen   Germany       NaN      NaN      Wilde (2003)  
10936            Sachsen   Germany       NaN      NaN      Wilde (2003)  
10937            Sachsen   Germany       NaN      NaN      Wilde (2003)  
10938            Sachsen   Germany       NaN      NaN      Wilde (2003)  
10939            Sachsen   Germany       NaN      NaN      Wilde (2003)  

[3417 rows x 12 columns]

In [29]:
# Group by year 
trials_germany_by_year = trials_germany.groupby([pd.Grouper(key='year')]).count()[['tried']].reset_index()

# rename the columns
trials_germany_by_year.columns = ['year', 'trials_count']

In [30]:
# line chart by altair
alt.Chart(trials_germany_by_year).mark_line().encode(
    x=alt.X('year'),
    y=alt.Y('trials_count'),
).properties(
    title={
        'text': 'Germany Witch Trials by year',
    }
)

/Users/direwolfsummer/.pyenv/versions/3.9.4/envs/j233-files-3.9.4/lib/python3.9/site-packages/altair/utils/core.py:317: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for col_name, dtype in df.dtypes.iteritems():


<VegaLite 4 object>

If you see this message, it means the renderer has not been properly enabled
for the frontend that you are using. For more information, see
https://altair-viz.github.io/user_guide/troubleshooting.html


![Germany Witch Trials by year](./img/germany_witch_trials_by_year.png)

In [32]:
# Group by gadm.adm1
trials_germany_by_region = trials_germany.groupby([pd.Grouper(key='gadm.adm1')]).count()[['tried']].reset_index()

# rename the columns
trials_germany_by_region.columns = ['region', 'trials_count']

In [33]:
# bar chart by altair
alt.Chart(trials_germany_by_region).mark_bar().encode(
    x=alt.X('region'),
    y=alt.Y('trials_count'),
).properties(
    title={
        'text': 'Germany Witch Trials by region',
    }
)

/Users/direwolfsummer/.pyenv/versions/3.9.4/envs/j233-files-3.9.4/lib/python3.9/site-packages/altair/utils/core.py:317: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for col_name, dtype in df.dtypes.iteritems():


<VegaLite 4 object>

If you see this message, it means the renderer has not been properly enabled
for the frontend that you are using. For more information, see
https://altair-viz.github.io/user_guide/troubleshooting.html


![Germany Witch Trials by region](./img/germany_witch_trials_by_region.png)

Germany was the Protestant Reformation’s birthplace and also the country most affected by Catholic–Protestant warfare. According to [Wikipedia's European wars of religion](https://en.wikipedia.org/wiki/European_wars_of_religion), in 1545, the Catholic Church began the Counter-Reformation against the growth of Protestantism. The conflicts culminated in the Thirty Years' War, which devastated Germany and killed one third of its population, a mortality rate twice that of World War I.

In 1487, two German Heinrich Kramer and Jacob Sprenger published Malleus Maleficarum – ‘Hammer of Witches’, which is probably the most influencial book on how to investigate accusations of witchcraft and execute witches. At that time, the Gutenberg printing press had just been invented along the Rhine River, and Kramer fully utilized it to print his book and spread his ideas. 

### Witch Trials in UK

In [34]:
trials_uk = trials_clean.loc[trials_clean['gadm.adm0']=='United Kingdom']
trials_uk

year  decade  century  tried  deaths city        gadm.adm2  \
2480   1699.0    1690     1600      1     NaN  NaN           Antrim   
2481   1711.0    1710     1700      8     NaN  NaN           Antrim   
2882   1569.0    1560     1500      1     0.0  NaN             Fife   
2883   1560.0    1560     1500      3     NaN  NaN              NaN   
2884   1562.0    1560     1500      1     NaN  NaN         Stirling   
...       ...     ...      ...    ...     ...  ...              ...   
10544  1676.0    1670     1600      1     1.0  NaN  Gloucestershire   
10545  1685.0    1680     1600      1     NaN  NaN        Wiltshire   
10546  1699.0    1690     1600      1     1.0  NaN            Essex   
10547  1732.0    1730     1700      1     NaN  NaN          Suffolk   
10548  1748.0    1740     1700      1     NaN  NaN          Suffolk   

              gadm.adm1       gadm.adm0  lon  lat          record.source  
2480   Northern Ireland  United Kingdom  NaN  NaN         Carlson (2013)  
2481   Northern Ireland  United Kingdom  NaN  NaN         Carlson (2013)  
2882           Scotland  United Kingdom  NaN  NaN  Goodare et al. (2003)  
2883           Scotland  United Kingdom  NaN  NaN  Goodare et al. (2003)  
2884           Scotland  United Kingdom  NaN  NaN  Goodare et al. (2003)  
...                 ...             ...  ...  ...                    ...  
10544           England  United Kingdom  NaN  NaN        Valletta (2000)  
10545           England  United Kingdom  NaN  NaN        Valletta (2000)  
10546           England  United Kingdom  NaN  NaN        Valletta (2000)  
10547           England  United Kingdom  NaN  NaN        Valletta (2000)  
10548           England  United Kingdom  NaN  NaN        Valletta (2000)  

[3750 rows x 12 columns]

In [35]:
# Group by year 
trials_uk_by_year = trials_uk.groupby([pd.Grouper(key='year')]).count()[['tried']].reset_index()

# rename the columns
trials_uk_by_year.columns = ['year', 'trials_count']

In [36]:
# line chart by altair
alt.Chart(trials_uk_by_year).mark_line().encode(
    x=alt.X('year'),
    y=alt.Y('trials_count'),
).properties(
    title={
        'text': 'UK Witch Trials by year',
    }
)

/Users/direwolfsummer/.pyenv/versions/3.9.4/envs/j233-files-3.9.4/lib/python3.9/site-packages/altair/utils/core.py:317: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for col_name, dtype in df.dtypes.iteritems():


<VegaLite 4 object>

If you see this message, it means the renderer has not been properly enabled
for the frontend that you are using. For more information, see
https://altair-viz.github.io/user_guide/troubleshooting.html


![UK Witch Trials by year](./img/uk_witch_trials_by_year.png)

According to [wikipedia](https://en.wikipedia.org/wiki/Witch_trials_in_England), witch trials are known to have occurred in England during the Middle Ages but cases are few. It was not until the second half of the 16th century that a widescale witch hunt took place. Witch hunt in the UK was as its most intense stage during the English Civil War (1642–1651) and the Puritan era of the mid-17th century. King James had shown a great interest in witch trials since the Copenhagen witch trials in 1589, and when he succeeded to the English throne in 1603, he sharpened the English Witchcraft Act the following year.

In [37]:
# Group by gadm.adm1
trials_uk_by_region = trials_uk.groupby([pd.Grouper(key='gadm.adm1')]).count()[['tried']].reset_index()

# rename the columns
trials_uk_by_region.columns = ['region', 'trials_count']

In [38]:
# bar chart by altair
alt.Chart(trials_uk_by_region).mark_bar().encode(
    x=alt.X('region'),
    y=alt.Y('trials_count'),
).properties(
    title={
        'text': 'UK Witch Trials by region',
    }
)

/Users/direwolfsummer/.pyenv/versions/3.9.4/envs/j233-files-3.9.4/lib/python3.9/site-packages/altair/utils/core.py:317: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for col_name, dtype in df.dtypes.iteritems():


<VegaLite 4 object>

If you see this message, it means the renderer has not been properly enabled
for the frontend that you are using. For more information, see
https://altair-viz.github.io/user_guide/troubleshooting.html


![UK Witch Trials by Region](./img/uk_witch_trials_by_region.png)

It seems that Scotland is the hotspot for witch trials in the UK. Let's take a closer look.

In [39]:
trials_scotland = trials_clean.loc[trials_clean['gadm.adm1']=='Scotland']
trials_scotland.sort_values(by=['tried'], ascending=False)

year  decade  century  tried  deaths city              gadm.adm2  \
4493  1649.0    1640     1600     48     NaN  NaN                    NaN   
5559  1661.0    1660     1600     15     NaN  NaN                    NaN   
4027  1649.0    1640     1600     15     NaN  NaN                    NaN   
4926  1650.0    1650     1600     12     NaN  NaN                    NaN   
4084  1649.0    1640     1600     12     NaN  NaN                    NaN   
...      ...     ...      ...    ...     ...  ...                    ...   
3935  1636.0    1630     1600      1     0.0  NaN           East Lothian   
3936  1636.0    1630     1600      1     0.0  NaN           East Lothian   
3937  1636.0    1630     1600      1     0.0  NaN  Dumfries and Galloway   
3938  1636.0    1630     1600      1     0.0  NaN               Highland   
6580  1479.0    1470     1400      1     1.0  NaN                    NaN   

     gadm.adm1       gadm.adm0  lon  lat          record.source  
4493  Scotland  United Kingdom  NaN  NaN  Goodare et al. (2003)  
5559  Scotland  United Kingdom  NaN  NaN  Goodare et al. (2003)  
4027  Scotland  United Kingdom  NaN  NaN  Goodare et al. (2003)  
4926  Scotland  United Kingdom  NaN  NaN  Goodare et al. (2003)  
4084  Scotland  United Kingdom  NaN  NaN  Goodare et al. (2003)  
...        ...             ...  ...  ...                    ...  
3935  Scotland  United Kingdom  NaN  NaN  Goodare et al. (2003)  
3936  Scotland  United Kingdom  NaN  NaN  Goodare et al. (2003)  
3937  Scotland  United Kingdom  NaN  NaN  Goodare et al. (2003)  
3938  Scotland  United Kingdom  NaN  NaN  Goodare et al. (2003)  
6580  Scotland  United Kingdom  NaN  NaN      Kieckhefer (1976)  

[3129 rows x 12 columns]

What happened in 1649?

Well, a series of witch trials happened during 1649–50 in Scotland, called The great Scottish witch hunt. It is one of five major hunts identified in early modern Scotland and it probably saw the most executions in a single year.

## Q3: Which country has the least witch trial cases in total?

In [40]:
# sort the dataframe by value
trials_by_country.sort_values(by='trials_count')

country  trials_count
2   Czech Republic             1
9          Ireland             4
14          Poland             9
0          Austria            16
4          Estonia            17
5          Finland            17
11      Luxembourg            20
13          Norway            20
8          Hungary            26
15           Spain            29
3          Denmark            90
10           Italy           107
12     Netherlands           314
16          Sweden           353
1          Belgium           671
6           France           807
17     Switzerland          1272
7          Germany          3417
18  United Kingdom          3750

It seems that Czech Republic, Ireland, and Poland are among the countries that saw the least witch trials activity during that time period.